### ISSUE TITLE
Unauthorized basic auth datasource requests trigger browser authentication

### ISSUE THREAD

**Issue by davkal** [grafana/grafana#15050]
0. Datasource that requires Basic auth, e.g., influx
1. Add dashboard
2. Add panel
3. Select influx datasource

Result:

<img width="541" alt="screenshot 2019-01-24 at 15 42 21" src="https://user-images.githubusercontent.com/859729/51685566-da7fd900-1fee-11e9-9843-38b8504abbc8.png">

On selection a datasource request happened that was returned with a 401.

**Comment by gowrisankar22** [comment]
@marefr  Any idea when this bug will be fixed?

**Comment by torkelo** [comment]
Not sure this is a bug, maybe user missed specify basic auth username and password datasource settings?

**Comment by gowrisankar22** [comment]
@torkelo Nope. User credentials were added already and moreover, I have logged in as an admin too.

**Comment by torkelo** [comment]
Can you show your

**Comment by torkelo** [comment]
Can you show your datasource settings ?

**Comment by gowrisankar22** [comment]
![image](https://user-images.githubusercontent.com/20590107/51960358-b56fe800-247e-11e9-92cf-9250f40a5cfb.png)

**Comment by gowrisankar22** [comment]
@torkelo  Nope I have tried with that already but after setting that also same error.
> t=2019-01-30T07:54:46+0000 lvl=info msg="Request Completed" logger=context userId=1 orgId=1 > uname=admin method=POST path=/api/datasources/proxy/173/api/v2/query status=401 ?
> remote_addr=1.1.1.1 time_ms=15 size=27 referer=https://grafana/datasources/edit/173

But I have the other data sources (ex:influxdb & azuremonitor) which works perfectly fine without checking anything.

**Comment by davkal** [comment]
@torkel This also happens with the influx docker block.

**Comment by gowrisankar22** [comment]
@davkal I am also using the docker influx(1.7.3) but issue didn't happen for me.

**Comment by davkal** [comment]
@gowrisankar22 Does your influx 1.7.3 instance have access restrictions? (Our docker test block has.)

**Comment by mikewop** [comment]
check your influxDB password. I had an issue with special characters in the password, so start with a simple password and make sure under the datasources configuration clicking the "Test & Save" button shows "Data source is working".

**Comment by walterdolce** [comment]
I am experiencing a similar problem with the Elasticsearch data source. I am using the latest version of Grafana at this time of writing (`6.2.4`). 

I have the log level set to `debug` but when I go on the data source pages and hit the "Test" button (the datasources can't be modified on the UI) I don't get any log and the HTTP basic auth popup appears. For another datasource, it doesn't. 

It's probably worth mentioning that by creating a data source entry manually with the same details used in the one that causes the HTTP basic auth to appear, the problem does not occur when I save or test the datasource.

@torkelo what's the best way to troubleshoot this? Where can I find the logs for the action I'm taking? (clicking on "Test", "Save and test", etc)

Thanks.

**Comment by walterdolce** [comment]
I think I found the culprit, for _my_ case at least. 

It looks like passwords containing a dollar sign (`$`) followed by any character cause problems. This used to work in `5.3.4` (version I upgraded from). Here's a few scenarios:
- `123123pass` - This works
- `+123123pass` - This works
- `+123123pass$` - This works
- `+123123pass$asdad` - This breaks

Hope this helps.

BTW, I think it would be good to produce more meaningful debug information - especially in [this area of the code](https://github.com/grafana/grafana/blob/2fff8f77dcdc90ab9a4890eeed95d8f3dced370b/pkg/tsdb/elasticsearch/client/client.go#L153). 

Having a `debug` mode that tells you `Creating new client`, `Executing multisearch` and `Executing request` is fine. But I don't think it's saying much. I believe logging what *actually* happens behind the scenes would greatly speed up the troubleshooting process when issues occur. I.e. I would log all requests and its parameters, headers, etc made to Elasticsearch (or any data source). I am conscious those requests could contain sensitive data (i.e. HTTP auth bearer tokens, etc) - I will leave to you to decide what's best :)

**Comment by coolriku** [comment]
> I think I found the culprit, for _my_ case at least.
> 
> It looks like passwords containing a dollar sign (`$`) followed by any character cause problems. This used to work in `5.3.4` (version I upgraded from). Here's a few scenarios:
> 
> * `123abc` - This works
> * `+123abc` - This works
> * `+123abc$` - This works
> * `+123abc$asdad` - This breaks
> * `+123abc$;asdad` - This **works** (note the semi colon `;`)
> 
> Hope this helps.
> 
> BTW, I think it would be good to produce more meaningful debug information - especially in [this area of the code](https://github.com/grafana/grafana/blob/2fff8f77dcdc90ab9a4890eeed95d8f3dced370b/pkg/tsdb/elasticsearch/client/client.go#L153).
> 
> Having a `debug` mode that tells you `Creating new client`, `Executing multisearch` and `Executing request` is fine. But I don't think it's saying much. I believe logging what _actually_ happens behind the scenes would greatly speed up the troubleshooting process when issues occur. I.e. I would log all requests and its parameters, headers, etc made to Elasticsearch (or any data source). I am conscious those requests could contain sensitive data (i.e. HTTP auth bearer tokens, etc) - I will leave to you to decide what's best :)
> 
> Had I had chance to see at the headers sent to Elasticsearch, I would have spotted the difference in the encoded HTTP auth credentials - which I suspect are being encoded incorrectly in this instance.

I had the same problem with a influxdb datasource.
When I changed the password to something without special characters it worked again.

**Comment by luqmanMohammed** [comment]
Grafana version: 6.2.0
Data source: InfluxDB
Same issue with my setup. Password has no special characters. Grafana is running as a container. Querying the InfluxDB Datasource from within the Grafana container using curl works, which ensure connectivity

**Comment by davkal** [comment]
I have just hit this with Loki in `master`.

**Comment by davkal** [comment]
Here is what's happening in Loki in Explore:

When Loki is loaded in the datasource selector, [it's checking the Loki version](https://github.com/grafana/grafana/blob/e0a2d4beacc379def0878655b3a3f7cceffbf3fc/public/app/plugins/datasource/loki/datasource.ts#L106) using the backend srv `datasourceRequest`, which is promise-yfied with a `catch`, but the browser auth triggers nonetheless. 

So even though the caller has a catch implemented, it's deferring the error handling to the browser. Any advice @hugohaggmark ?

**Comment by hugohaggmark** [comment]
@davkal will have a look tomorrow, thanks for ping!

**Comment by torkelo** [comment]
> So even though the caller has a catch implemented, it's deferring the error handling to the browser.

Not sure what this means? is the browser handling the error, how?

**Comment by davkal** [comment]
It shows a basic auth form:

![Screenshot 2020-03-24 at 17 08 45](https://user-images.githubusercontent.com/859729/77449450-3a8c2e00-6df2-11ea-830f-8d10a9ccecf2.png)

**Comment by torkelo** [comment]
Oh, yea think 401 plus basic auth Authorization response header will do that. Not sure if there is a way in the frontend to prevent it, maybe

**Comment by hugohaggmark** [comment]
I looked into this briefly and it seems like BackendSrv works as designed. You probably see the pop-up appear several times, right? I tried to catch the 401 error without luck in BackendSrv which shows that there's some browser specific functionality taking over.

So I don't think we can prevent the Browser from showing the pop-up with failing Basic or Digest Authentication. Your suggestion @davkal to use another response code would be one way forward to solve this.

**Comment by DanCech** [comment]
Assuming the request is being proxied through the server we could blanket rewrite 401 into a 400 response in the proxy, since we know that a 401 isn't going to be helpful to the end user in Grafana.

**Comment by davkal** [comment]
Wouldn't it be enough to remove the `Www-Authenticate` header? 401s have been around for a while and rarely brought up the Basic auth popup. Did grafana cloud add this header recently?
```
Www-Authenticate: BASIC realm="For access to grafanacloud, provide instance id and api key."
```

**Comment by DanCech** [comment]
Technically www-authenticate is required by the rfc for 401, though Grafana could substitute its own with a custom scheme (also technically not rfc-compliant since altering the www-authenticate header in a proxy is not allowed, though you could argue there that the grafana backend is not a proxy as-such and is just part of the application).

https://tools.ietf.org/html/rfc7235#section-3.1

**Comment by DanCech** [comment]
For direct mode the only options would be to find out if there is a way to intercept the 401 before the user prompt, or update the source itself to not return a 401 in that case.

For requests going via the proxy another option would be to update the backend and the client so that any response code from the source is passed in a custom header and the backend always returns a 200 if it was able to proxy the request successfully, then the client is responsible for checking the header to find out what the original response code was.

**Comment by Rohlik** [comment]
I hit on this problem with Grafana 9.1.5 and Prometheus datasource with basicAuth 😞.
Password like `2aty&2Rgg4@njGmh$XFg` caused the issue. When I password changed to something different (without `$`), then everything works fine.

**Comment by mathijswesterhof** [comment]
I'm currently setting up Grafana 9.4.3 stable (amd64) and I got this issue with Prometheus my password does not contain any special symbols and logging in with that password on the web interface of Prometheus works. When i configure the Datasource the popup shows everything is okey but when I try to configure a dashboard nothing shows up and I can see a 401 in the console. I also spotted a message in the info and debug log.  It pops up once and when you cancel it the 401 shows. I tried configuring a Grafana user with the same username and password and then the whole thing goes into an infinite loop respawning the login on the dashboard page even when i try to cancel. 
Is there a http-header that I'm missing or a config?  (disabling prometheus simple auth directly solves the issue, and I got the basic auth settings in grafana correct because if i fill in some random password it fails with a popup)
